datafusion-examples/examples/struct_field_rewrite.rs

```diff
@@ -0,0 +1,353 @@
+// expressions like `struct_col['a'] = 10` to use a flattened column name
+// `_struct_col.a` if it exists in the file schema.
+#[tokio::main]
+async fn main() -> Result<()> {
+    println!("=== Creating example data with structs and flattened fields ===");
+    println!("NOTE: This example demonstrates filter expression rewriting for struct field access.");
+    println!("      We deliberately create different values in the struct field vs. flattened column");
+    println!("      to prove that the rewriting system is actually using the flattened column.");
+
+    // Create sample data with both struct columns and flattened fields
+    let (table_schema, batch) = create_sample_data();
+
+    let store = InMemory::new();
+    let buf = {
+        let mut buf = vec![];
+
+        let props = WriterProperties::builder()
+            .set_max_row_group_size(1)
+            .build();
+
+        let mut writer = ArrowWriter::try_new(&mut buf, batch.schema(), Some(props))
+            .expect("creating writer");
+
+        writer.write(&batch).expect("Writing batch");
+        writer.close().unwrap();
+        buf
+    };
+    let path = Path::from("example.parquet");
+    let payload = PutPayload::from_bytes(buf.into());
+    store.put(&path, payload).await?;
+
+    // Create a custom table provider that rewrites struct field access
+    let table_provider = Arc::new(ExampleTableProvider::new(table_schema));
+
+    // Set up query execution
+    let ctx = SessionContext::new();
+
+    // Register our table
+    ctx.register_table("structs", table_provider)?;
+
+    ctx.runtime_env().register_object_store(
+        ObjectStoreUrl::parse("memory://")?.as_ref(),
+        Arc::new(store),
+    );
+
+    println!("\n=== Showing all data ===");
+    let batches = ctx.sql("SELECT * FROM structs").await?.collect().await?;
+    arrow::util::pretty::print_batches(&batches)?;
+
+    println!("\n=== Running query with struct field access and filter < 30 ===");
+    println!("Query: SELECT user_info['name'] FROM structs WHERE user_info['age'] < 30");
+
+    let batches = ctx
+        .sql("SELECT * FROM structs WHERE user_info['age'] > 30")
+        .await?
+        .collect()
+        .await?;
+
+    #[rustfmt::skip]
+    let expected = vec![
+        "+--------------------------+",
+        "| user_info                |",
+        "+--------------------------+",
+        "| {name: Charlie, age: 35} |",
+        "+--------------------------+",
+    ];
+    assert_batches_eq!(expected, &batches);
+
+    println!("\n=== Running query with struct field access and filter > 130 ===");
+    println!("Query: SELECT user_info['name'] FROM structs WHERE user_info['age'] < 30");
+
+    let batches = ctx
+        .sql("explain analyze SELECT user_info['name'] FROM structs WHERE user_info['age'] < 30")
+        .await?
+        .collect()
+        .await?;
+    arrow::util::pretty::print_batches(&batches)?;
```

Comment:
Need a better assertion here but the point is that it shows `row_groups_pruned_statistics=2` (without the rewrite it's `row_groups_pruned_statistics=0`).